src/microsAuthentication/content.md

```diff
@@ -0,0 +1,1157 @@
+
+All authorisation servers have to do is to:
+
+1. **Authenticate the caller** - The caller should have a valid and verifiable identity.
+1. **Generate a token with a limited scope, validity and the desired audience.**
+1. **Validate a token** - Service to service communication is allowed only if the token is legit for the two services involved.
+
+Examples of Authorisation servers are tools such as [Keycloak](https://www.keycloak.org/) or [Dex](https://github.com/dexidp/dex).
+
+You might have not noticed, but even Kubernetes offers the same primitives as the authorisation service with Service Account, Roles and RoleBindings.
+
+In Kubernetes, you assign identities using Service Accounts.
+
+Service accounts are then linked to roles to that grant access to resources.
+
+Tokens are generated as soon as you create the Service Account and stored in a Secret.
+
+Users and Pods can use the secret as a mechanism to authenticate to the API and make requests.
+
+You only receive successful replies for the resources you are entitled to consume.
+
+If your Role grant you access to create and delete Pods, you won't be able to amend Secrets, or create ConfigMaps — for example.
+
+_Could you use Service Accounts as a mechanism to authenticate requests between apps in the cluster?_
+
+Kubernetes offers the same primitives to authentication requests just like Keycloak and Dex.
+
+_What if the Kubernetes API could be used as an Authorisation server?_
+
+_Could you use Service Accounts as a mechanism to authenticate requests between apps in the cluster?_
+
+Let's try that.
+
+You will now deploy two services:
+
+1. You will refer to these services as the API service (API) and the Secret store.
+1. They are written in the Go programming language and they communicate via HTTP.
+1. Each service runs in its own namespaces and use dedicated service accounts.
+1. The secret store requires that any client identifies itself before making a request.
+
+## Creating the cluster
+
+You will need access to a Kubernetes cluster with the [ServiceAccountVolume projection](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection) feature enabled.
+
+Don't worry if you don't know what a ServiceAccountVolume is — you will learn more about it later on in the article.
+
+The ServiceAccountVolume projection requires the Kubernetes API server to be started with certain specific API flags.
+
+The support for different managed Kubernetes providers may vary.
+
+However, you can enable it in [minikube](https://github.com/kubernetes/minikube) with:
+
+```terminal|command=1-6|title=bash
+minikube start \
+  --extra-config=apiserver.service-account-signing-key-file=\
+    /var/lib/minikube/certs/sa.key \
+  --extra-config=apiserver.service-account-issuer=\
+    kubernetes/serviceaccount \
+  --extra-config=apiserver.service-account-api-audiences=api
+```
```

Comment:
I don't have `/var/lib/minikube/`. My minikube fails:

```bash
minikube start \
  --extra-config=apiserver.service-account-signing-key-file=\
    /var/lib/minikube/certs/sa.key \
  --extra-config=apiserver.service-account-issuer=\
    kubernetes/serviceaccount \
  --extra-config=apiserver.service-account-api-audiences=api
😄  minikube v1.14.0 on Darwin 10.15.7
✨  Automatically selected the hyperkit driver
👍  Starting control plane node minikube in cluster minikube
🔥  Creating hyperkit VM (CPUs=2, Memory=4000MB, Disk=20000MB) ...
🐳  Preparing Kubernetes v1.19.2 on Docker 19.03.12 ...
    ▪ apiserver.service-account-signing-key-file=
    ▪ apiserver.service-account-issuer=
    ▪ apiserver.service-account-api-audiences=api
💢  initialization failed, will try again: run: /bin/bash -c "sudo env PATH=/var/lib/minikube/binaries/v1.19.2:$PATH kubeadm init --config /var/tmp/minikube/kubeadm.yaml  --ignore-preflight-errors=DirAvailable--etc-kubernetes-manifests,DirAvailable--var-lib-minikube,DirAvailable--var-lib-minikube-etcd,FileAvailable--etc-kubernetes-manifests-kube-scheduler.yaml,FileAvailable--etc-kubernetes-manifests-kube-apiserver.yaml,FileAvailable--etc-kubernetes-manifests-kube-controller-manager.yaml,FileAvailable--etc-kubernetes-manifests-etcd.yaml,Port-10250,Swap": Process exited with status 1
stdout:
[init] Using Kubernetes version: v1.19.2
[preflight] Running pre-flight checks
[preflight] Pulling images required for setting up a Kubernetes cluster
[preflight] This might take a minute or two, depending on the speed of your internet connection
[preflight] You can also perform this action in beforehand using 'kubeadm config images pull'
[certs] Using certificateDir folder "/var/lib/minikube/certs"
[certs] Using existing ca certificate authority
[certs] Using existing apiserver certificate and key on disk
[certs] Generating "apiserver-kubelet-client" certificate and key
[certs] Generating "front-proxy-ca" certificate and key
[certs] Generating "front-proxy-client" certificate and key
[certs] Generating "etcd/ca" certificate and key
[certs] Generating "etcd/server" certificate and key
[certs] etcd/server serving cert is signed for DNS names [localhost minikube] and IPs [192.168.64.21 127.0.0.1 ::1]
[certs] Generating "etcd/peer" certificate and key
[certs] etcd/peer serving cert is signed for DNS names [localhost minikube] and IPs [192.168.64.21 127.0.0.1 ::1]
[certs] Generating "etcd/healthcheck-client" certificate and key
[certs] Generating "apiserver-etcd-client" certificate and key
[certs] Generating "sa" key and public key
[kubeconfig] Using kubeconfig folder "/etc/kubernetes"
[kubeconfig] Writing "admin.conf" kubeconfig file
[kubeconfig] Writing "kubelet.conf" kubeconfig file
[kubeconfig] Writing "controller-manager.conf" kubeconfig file
[kubeconfig] Writing "scheduler.conf" kubeconfig file
[kubelet-start] Writing kubelet environment file with flags to file "/var/lib/kubelet/kubeadm-flags.env"
[kubelet-start] Writing kubelet configuration to file "/var/lib/kubelet/config.yaml"
[kubelet-start] Starting the kubelet
[control-plane] Using manifest folder "/etc/kubernetes/manifests"
[control-plane] Creating static Pod manifest for "kube-apiserver"
[control-plane] Creating static Pod manifest for "kube-controller-manager"
[control-plane] Creating static Pod manifest for "kube-scheduler"
[etcd] Creating static Pod manifest for local etcd in "/etc/kubernetes/manifests"
[wait-control-plane] Waiting for the kubelet to boot up the control plane as static Pods from directory "/etc/kubernetes/manifests". This can take up to 4m0s
[kubelet-check] Initial timeout of 40s passed.

	Unfortunately, an error has occurred:
		timed out waiting for the condition

	This error is likely caused by:
		- The kubelet is not running
		- The kubelet is unhealthy due to a misconfiguration of the node in some way (required cgroups disabled)

	If you are on a systemd-powered system, you can try to troubleshoot the error with the following commands:
		- 'systemctl status kubelet'
		- 'journalctl -xeu kubelet'

	Additionally, a control plane component may have crashed or exited when started by the container runtime.
	To troubleshoot, list all containers using your preferred container runtimes CLI.

	Here is one example how you may list all Kubernetes containers running in docker:
		- 'docker ps -a | grep kube | grep -v pause'
		Once you have found the failing container, you can inspect its logs with:
		- 'docker logs CONTAINERID'


stderr:
W1120 13:02:14.574236    2360 configset.go:348] WARNING: kubeadm cannot validate component configs for API groups [kubelet.config.k8s.io kubeproxy.config.k8s.io]
	[WARNING Service-Kubelet]: kubelet service is not enabled, please run 'systemctl enable kubelet.service'
error execution phase wait-control-plane: couldn't initialize a Kubernetes cluster
To see the stack trace of this error execute with --v=5 or higher
```

Reply:
Not sure, can you please try with the latest minikube version? I just tried now and it works for me:
```
$ minikube version
minikube version: v1.15.1
commit: 23f40a012abb52eff365ff99a709501a61ac5876
```

My script:

```
% cat start_minikube.sh 
#!/bin/bash

minikube start --extra-config=apiserver.service-account-signing-key-file=/var/lib/minikube/certs/sa.key --extra-config=apiserver.service-account-issuer=kubernetes/serviceaccount --extra-config=apiserver.service-account-api-audiences=api
```

Reply:
```bash
$ ls /var/lib/minikube/
ls: /var/lib/minikube/: No such file or directory
```

What's your like?

Reply:
Ok, the certs form me are in `~/.minikube/certs`.
But I don't have a `sa.cert` only a `ca.cert`. Did you generate the cert yourself?